ami/models/components/unet/unet.py

```diff
@@ -0,0 +1,204 @@
+# Ref: https://github.com/facebookresearch/RCDM
+
+import math
+
+import torch
+import torch.nn as nn
+
+from .decoder_block import UnetDecoderBlock
+from .encoder_block import UnetEncoderBlock
+from .middle_block import UnetMiddleBlock
+
+
+def timestep_embedding(timesteps: torch.Tensor, dim: int, max_period: int = 10000) -> torch.Tensor:
+    """Create sinusoidal timestep embeddings.
+
+    Args:
+        timesteps (torch.Tensor):
+            Timesteps for noise reduction.
+            (shape: [batch_size])
+        dims (int):
+            Output dims.
+        max_period (int):
+            Controls the minimum frequency of the embeddings.
+    Returns:
+        torch.Tensor:
+            Embedded timesteps.
+            (shape: [batch_size, dim])
+    """
+    assert dim % 2 == 0
+    half = dim // 2
+    freqs = torch.exp(-math.log(max_period) * torch.arange(start=0, end=half, dtype=torch.float) / half).to(
+        device=timesteps.device
+    )
+    args = timesteps[:, None].float() * freqs[None]
+    embedding = torch.cat([torch.cos(args), torch.sin(args)], dim=-1)
+    return embedding
+
+
+class UNet(nn.Module):
+    """Unet architecture for noise prediction from input images."""
+
+    def __init__(
+        self,
+        in_channels: int,
+        timestep_hidden_dim: int,
+        timestep_embed_dim: int,
+        out_channels: int,
+        encoder_blocks_in_and_out_channels: list[tuple[int, int]],
+        n_res_blocks: int,
+        attention_start_depth: int,
+        num_heads: int,
+        ssl_latent_dim: int,
+    ) -> None:
+        """Unet architecture for noise prediction from input images.
+
+        Args:
+            in_channels (int):
+                Input images num of channels.
+            timestep_hidden_dim (int):
+                Hidden dims of timestep embedding.
+            timestep_embed_dim (int):
+                Dims of timestep embedding.
+            out_channels (int):
+                Output images num of channels.
+            encoder_blocks_in_and_out_channels (list[tuple[int, int]]):
+                Specify (in_channels, out_channels) of each encoder block,
+                in order from the shallowest block to the deepest block.
+            n_res_blocks (int):
+                num of resblocks in each encoder block.
+            attention_start_depth (int):
+                Append attention layers for blocks deeper than attention_start_depth.
```

Comment:
0オリジンであることは明記した方がいい。
```suggestion
            attention_start_depth (int):
                Append attention layers for blocks deeper than attention_start_depth. 0 origin.
```